jax/_src/lru_cache.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import logging
+import os
+import tensorflow as tf
+from typing import Any
+
+from jax._src import config
+from jax._src.compilation_cache_interface import CacheInterface
+from jax._src.gfile_cache import GFileCacheImpl
+
+
+filelock: Any = None
+try:
+  import filelock
+except ImportError:
+  pass
+
+
+logger = logging.getLogger(__name__)
+
+
+def is_path_gs(path: str) -> bool:
+    return path.startswith("gs://")
+
+
+def get_size(path: str) -> int:
+    return tf.io.gfile.stat(path).length
+
+
+def get_mtime(path: str) -> int:
+    return tf.io.gfile.stat(path).mtime_nsec
+
+
+def update_mtime(path: str) -> None:
+    if not is_path_gs(path):
+        os.utime(path)  # set mtime (and also atime) to current time
+    else:
+        tmp_path = f"{path}.tmp"
+        tf.io.gfile.rename(path, tmp_path)
+        tf.io.gfile.rename(tmp_path, path)
+
+
+class LRUCacheImpl:
+
+    def __init__(self, path: str, *, max_size: int, timeout=10):
+        if filelock is None:
+            raise RuntimeError("Please install filelock to use the LRUCache")
+
+        self.path = path
+        self.max_size = max_size
+        os.makedirs(path, exist_ok=True)
+        self.timeout = timeout
+
+        self.lock_file = os.path.join(self.path, ".lockfile")
```

Comment:
nit: I would call this `lock_path`, to make clear it's the filepath and not the file object itself

Reply:
Done